server/src/main/java/org/elasticsearch/common/unit/Processors.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.unit;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.Locale;
+import java.util.Objects;
+
+import static java.lang.String.format;
+
+public class Processors implements Writeable, Comparable<Processors>, ToXContentFragment {
+    public static final Processors ZERO = new Processors(0.0);
+    public static final Processors MAX_PROCESSORS = new Processors(Double.MAX_VALUE);
+
+    public static final Version FLOAT_PROCESSORS_SUPPORT_VERSION = Version.V_8_3_0;
+    public static final Version DOUBLE_PROCESSORS_SUPPORT_VERSION = Version.V_8_5_0;
+    static final int NUMBER_OF_DECIMAL_PLACES = 5;
+    private static final double MIN_REPRESENTABLE_PROCESSORS = 1E-5;
+
+    private final double count;
+
+    public Processors(double count) {
+        if (validNumberOfProcessors(count) == false) {
+            throw new IllegalArgumentException("processors must be a non-negative number; provided [" + count + "]");
+        }
+
+        // Avoid rounding up to MIN_REPRESENTABLE_PROCESSORS when 0 processors are used
+        if (count == 0.0) {
+            this.count = count;
+        } else {
+            this.count = Math.max(
+                MIN_REPRESENTABLE_PROCESSORS,
+                new BigDecimal(count).setScale(NUMBER_OF_DECIMAL_PLACES, RoundingMode.HALF_UP).doubleValue()
+            );
+        }
+    }
+
+    @Nullable
+    public static Processors of(Double count) {
```

Comment:
I wonder if we should move all `Processors` construction to use an `of` method, just to simplify and avoid mistakes? That `of` method could hold the validation too then, helping on the `ZERO` part I commented on in another comment (though a minor point).